src/rebuild/srv.c

```diff
@@ -320,14 +337,109 @@ rpt_lookup(uuid_t pool_uuid, uint32_t opc, unsigned int ver, unsigned int gen)
 	return found;
 }
 
+static void
+update_and_warn_for_slow_engines(struct rebuild_global_pool_tracker *rgt)
+{
+	int    i;
+	int    scan_ct = 0;
+	int    pull_ct = 0;
+	int    done_ct;
+	int    wait_ct;
+	bool   scan_gl = false;
+	bool   pull_gl = false;
+	bool   do_warn = false;
+	double now     = ABT_get_wtime();
+	double tw      = now - rgt->rgt_last_warn_ts; /* time since last warning logged */
+	bool   warned  = false;
+
+	/* Throttle warnings to not more often than once per 2 minutes */
+	do_warn = (tw >= 120);
+
+	/* Count scan/pull progress and warn for any ranks that haven't provided updates recently */
+	for (i = 0; i < rgt->rgt_servers_number; i++) {
+		double   tu = now - rgt->rgt_servers[i].last_update;
+		d_rank_t r  = rgt->rgt_servers[i].rank;
+
+		if (rgt->rgt_servers[i].scan_done) {
+			scan_ct++;
+			if (rgt->rgt_servers[i].pull_done) {
+				pull_ct++;
+				continue;
+			}
+		}
+
+		if (!do_warn)
+			continue;
+
+		if (tu > 30) {
+			D_WARN(DF_RB ": no updates from rank %u in %8.3f seconds. "
+				     "scan_done=%d pull_done=%d\n",
+			       DP_RB_RGT(rgt), r, tu, rgt->rgt_servers[i].scan_done,
+			       rgt->rgt_servers[i].pull_done);
+			warned = true;
+		}
+	}
+
+	scan_gl = (scan_ct == rgt->rgt_servers_number);
+	pull_gl = (pull_ct == rgt->rgt_servers_number);
+	if (scan_gl && pull_gl)
+		return;
+
+	/* Determine if scan/pull progress is almost done; mark the time and warn on possible stall.
+	 */
+	done_ct = scan_gl ? pull_ct : scan_ct;
+	wait_ct = rgt->rgt_servers_number - done_ct;
+#if 0
```

Comment:
oops, this should be removed. Let me do this cleanup in a future patch, since we have a lot invested in length CI waiting time with this change.